File: mobile/ios/Features/Settings/SettingsViewModel.swift
import Observation
import SwiftUI

@MainActor
@Observable
final class SettingsViewModel {
    private let api: APIClient

    var bundle: SettingsBundle?
    var isLoading = false
    var loadError: String?
    var serverRequiresUpgrade = false

    init(api: APIClient) {
        self.api = api
    }

    func load() async {
        isLoading = true
        defer { isLoading = false }
        do {
            bundle = try await api.settings()
            loadError = nil
            serverRequiresUpgrade = false
        } catch APIError.server(let status, _) where status == 404 {
            // 老版本服务端尚未注册 /settings/me；不要把它误报成网络故障。
            loadError = nil
            serverRequiresUpgrade = true
        } catch {
            // 加载失败不显示默认值冒充服务端值。
            loadError = error.localizedDescription
            serverRequiresUpgrade = false
        }
    }

    // MARK: 识别规则

    /// 乐观更新 + 失败回滚：立即改本地，失败时恢复旧值并抛错给 UI 提示。
    func saveDetection(keywords: [String], aiSemanticsEnabled: Bool) async throws {
        guard var current = bundle else { return }
        let previous = current.detection
        current.detection = DetectionSettings(keywords: keywords, aiSemanticsEnabled: aiSemanticsEnabled)
        bundle = current
        do {
            let saved = try await api.updateDetectionSettings(
                DetectionSettingsUpdate(keywords: keywords, aiSemanticsEnabled: aiSemanticsEnabled)
            )
            bundle?.detection = saved
        } catch {
            bundle?.detection = previous
            throw error
        }
    }

    // MARK: 工作时间

    func saveWorkSchedule(_ schedule: WorkSchedule) async throws {
        guard let current = bundle else { return }
        let previous = current.workSchedule
        bundle?.workSchedule = schedule
        do {
            let saved = try await api.updateWorkSchedule(
                WorkScheduleUpdate(
                    timezone: schedule.timezone,
                    slots: schedule.slots,
                    autoReplyOutsideHours: schedule.autoReplyOutsideHours
                )
            )
            bundle?.workSchedule = saved
        } catch {
            bundle?.workSchedule = previous
            throw error
        }
    }

    // MARK: 通知

    func saveNotifications(_ prefs: NotificationSettings) async throws {
        guard let current = bundle else { return }
        let previous = current.notifications
        bundle?.notifications = prefs
        do {
            let saved = try await api.updateNotificationSettings(
                NotificationSettingsUpdate(
                    newOpportunityEnabled: prefs.newOpportunityEnabled,
                    aiRepliedEnabled: prefs.aiRepliedEnabled,
                    dailyDigestEnabled: prefs.dailyDigestEnabled,
                    urgentOnly: prefs.urgentOnly
                )
            )
            bundle?.notifications = saved
        } catch {
            bundle?.notifications = previous
            throw error
        }
    }
}

File: mobile/android/app/src/main/kotlin/com/codeiy/im/feature/settings/SettingsViewModel.kt
package com.codeiy.im.feature.settings

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.codeiy.im.core.network.RadarApi
import com.codeiy.im.core.network.api
import com.codeiy.im.model.DetectionSettings
import com.codeiy.im.model.DetectionSettingsUpdate
import com.codeiy.im.model.NotificationSettings
import com.codeiy.im.model.NotificationSettingsUpdate
import com.codeiy.im.model.SettingsBundle
import com.codeiy.im.model.WorkSchedule
import com.codeiy.im.model.WorkScheduleSlotDTO
import com.codeiy.im.model.WorkScheduleUpdate
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch

class SettingsViewModel(private val service: RadarApi) : ViewModel() {
    data class UiState(
        val bundle: SettingsBundle? = null,
        val isLoading: Boolean = false,
        val loadError: String? = null,
    )

    private val _state = MutableStateFlow(UiState())
    val state: StateFlow<UiState> = _state

    fun load() {
        _state.value = _state.value.copy(isLoading = true)
        viewModelScope.launch {
            try {
                _state.value = _state.value.copy(bundle = api { service.settings() }, isLoading = false, loadError = null)
            } catch (e: Exception) {
                // 加载失败不显示默认值冒充服务端值。
                _state.value = _state.value.copy(isLoading = false, loadError = e.message)
            }
        }
    }

    /** 乐观更新 + 失败回滚：立即改本地，失败恢复旧值并回调错误。 */
    fun saveDetection(keywords: List<String>, aiSemanticsEnabled: Boolean, onError: (String) -> Unit) {
        val bundle = _state.value.bundle ?: return
        val previous = bundle.detection
        _state.value = _state.value.copy(bundle = bundle.copy(detection = DetectionSettings(keywords, aiSemanticsEnabled)))
        viewModelScope.launch {
            try {
                val saved = api { service.updateDetection(DetectionSettingsUpdate(keywords, aiSemanticsEnabled)) }
                _state.value = _state.value.copy(bundle = _state.value.bundle?.copy(detection = saved))
            } catch (e: Exception) {
                _state.value = _state.value.copy(bundle = _state.value.bundle?.copy(detection = previous))
                onError(e.message ?: "保存失败")
            }
        }
    }

    fun saveWorkSchedule(
        timezone: String,
        slots: List<WorkScheduleSlotDTO>,
        autoReplyOutsideHours: Boolean,
        onError: (String) -> Unit,
    ) {
        val bundle = _state.value.bundle ?: return
        val previous = bundle.workSchedule
        _state.value = _state.value.copy(
            bundle = bundle.copy(
                workSchedule = WorkSchedule(timezone, slots, autoReplyOutsideHours, false)
            )
        )
        viewModelScope.launch {
            try {
                val saved = api {
                    service.updateWorkSchedule(
                        WorkScheduleUpdate(timezone, slots, autoReplyOutsideHours)
                    )
                }
                _state.value = _state.value.copy(bundle = _state.value.bundle?.copy(workSchedule = saved))
            } catch (e: Exception) {
                _state.value = _state.value.copy(bundle = _state.value.bundle?.copy(workSchedule = previous))
                onError(e.message ?: "保存失败")
            }
        }
    }

    fun saveNotifications(prefs: NotificationSettings, onError: (String) -> Unit) {
        val bundle = _state.value.bundle ?: return
        val previous = bundle.notifications
        _state.value = _state.value.copy(bundle = bundle.copy(notifications = prefs))
        viewModelScope.launch {
            try {
                val saved = api {
                    service.updateNotifications(
                        NotificationSettingsUpdate(
                            prefs.newOpportunityEnabled,
                            prefs.aiRepliedEnabled,
                            prefs.dailyDigestEnabled,
                            prefs.urgentOnly,
                        )
                    )
                }
                _state.value = _state.value.copy(bundle = _state.value.bundle?.copy(notifications = saved))
            } catch (e: Exception) {
                _state.value = _state.value.copy(bundle = _state.value.bundle?.copy(notifications = previous))
                onError(e.message ?: "保存失败")
            }
        }
    }
}
